packages/gatsby/src/bootstrap/get-config-file.ts

```diff
@@ -14,114 +15,186 @@ export async function getConfigFile(
+    report.panic({
+      id: `11902`,
+      error: error,
+      context: {
+        configName,
+        message: error.message,
+      },
+    })
+  }
+
+  return compiledResult
+}
+
+async function attemptImportUncompiled(
+  siteDirectory: string,
+  configName: string,
+  distance: number
+): Promise<{
+  configModule: unknown
+  configFilePath: string
+}> {
+  let uncompiledResult
+
+  const uncompiledConfigPath = path.join(siteDirectory, configName)
 
-  // Attempt to find compiled gatsby-config.js in .cache/compiled/gatsby-config.js
   try {
-    configPath = path.join(`${siteDirectory}/${COMPILED_CACHE_DIR}`, configName)
-    configFilePath = require.resolve(configPath)
-    configModule = require(configFilePath)
-  } catch (outerError) {
-    // Not all plugins will have a compiled file, so the err.message can look like this:
-    // "Cannot find module '<root>/node_modules/gatsby-source-filesystem/.cache/compiled/gatsby-config'"
-    // But the compiled file can also have an error like this:
-    // "Cannot find module 'foobar'"
-    // So this is trying to differentiate between an error we're fine ignoring and an error that we should throw
-    const isModuleNotFoundError = outerError.code === `MODULE_NOT_FOUND`
-    const isThisFileRequireError =
-      outerError?.requireStack?.[0]?.includes(`get-config-file`) ?? true
-
-    // User's module require error inside gatsby-config.js
-    if (!(isModuleNotFoundError && isThisFileRequireError)) {
+    uncompiledResult = await attemptImport(siteDirectory, uncompiledConfigPath)
+  } catch (error) {
+    if (!testImportError(uncompiledConfigPath, error)) {
       report.panic({
-        id: `11902`,
-        error: outerError,
+        id: `10123`,
+        error,
         context: {
           configName,
-          message: outerError.message,
+          message: error.message,
         },
       })
     }
+  }
+
+  if (uncompiledResult?.configModule && uncompiledResult?.configFilePath) {
+    return uncompiledResult
+  }
+
+  const error = new Error(`Cannot find package '${uncompiledConfigPath}'`)
+
+  const { tsConfig, nearMatch } = await checkTsAndNearMatch(
+    siteDirectory,
+    configName,
+    distance
+  )
+
+  // gatsby-config.ts exists but compiled gatsby-config.js does not
+  if (tsConfig) {
+    report.panic({
+      id: `10127`,
+      error,
+      context: {
+        configName,
+      },
+    })
+  }
+
+  // gatsby-config is misnamed
+  if (nearMatch) {
+    const isTSX = nearMatch.endsWith(`.tsx`)
+    report.panic({
+      id: `10124`,
+      error,
+      context: {
+        configName,
+        nearMatch,
+        isTSX,
+      },
+    })
+  }
+
+  // gatsby-config.js/gatsby-config.mjs is incorrectly located in src/gatsby-config.js
+  const srcDirectoryHasJsConfig = existsSync(
+    path.join(siteDirectory, `src`, configName + `.js`)
+  )
+  const srcDirectoryHasMJsConfig = existsSync(
+    path.join(siteDirectory, `src`, configName + `.mjs`)
+  )
```

Comment:
@marvinjude Couldn't you have used the `resolveConfigFilePath` utility for this?

Reply:
Actually, `resolveConfigFilePath` does some additional checks above to warn on multiple configs which we won't need here since we're just trying to see if they have config incorrectly located in `src` directory

Reply:
Then that codepath won't be hit - that's totally fine 👍 We should use `resolveConfigFilePath` because right now you'd be duplicating logic

Reply:
The calls to `fs` are cached and since they'd happen anyways there are no wasted calls

Reply:
I don't think we'll hit that edge-case but we could add an optional third arg (`true` by default) that enables the warning (and then for this src/gatsby-config instance it'll be `false`)

Reply:
Agree we can probably use the same util, I'll add a task to the gatsby-config.ts ticket to circle back to this

Reply:
Went ahead and did this in https://github.com/gatsbyjs/gatsby/pull/37068/commits/ac899b30d0d6fe8ce7d8d4a7387f67c422d195ad